### DISCUSSION TITLE
Forcing nuget.org in nuget provider

### DISCUSSION THREAD

**Discussion by mandalorianbob** [cake-build/cake#3726] [Q&A] [answered]
The documentation for the Nuget Provider at https://cakebuild.net/docs/writing-builds/preprocessor-directives/tool/nuget-provider says 
>  By default, the provider will attempt to install tools from nuget.org.

But it apparently actually uses the providers that are specified in the local nuget config, because it keeps giving me errors about
> Unable to load the service index for source [**URI to private nuget feed hosted on AzDO**].

I tried
``` 
.InstallTool(new Uri("nuget:https://www.nuget.org?package=7-Zip.CommandLine"))
```
But got
> An error occurred while retrieving package metadata for '7-Zip.CommandLine' from source 'https://www.nuget.org/'

So I'm wondering how I can force the nuget provider to use nuget.org.

**Comment by gitfool** [comment]
Besides any support in Cake itself, would the [source mapping](https://devblogs.microsoft.com/nuget/announcing-nuget-6/#source-mapping) feature of NuGet 6.0 be a better way to enforce such desired behavior?

**Reply by mandalorianbob** [reply]
I like the idea of that, and it would probably work in a single Cake project. But I'm actually distributing my Cake tasks to my team via the same NuGet private feed, which we've actually set up to be a downstream feed from Nuget.org.

The source mapping looks like it requires a nuget.config. If I could somehow specify source mapping in my library/nuget package, it'd probably be fine. But since it doesn't look like I can, it means I'd have to have my team set up a special nuget.config for the build projects they create that use my library. 

That's a pretty good workaround, but I'd rather my library be self-contained. So it'd work a lot better for me if it actually did just preference nuget.org.

**Reply by mandalorianbob** [reply]
Oh actually I can't use this workaround at the moment. I can't mix private and public feeds. I had forgotten about that.

**Comment by pascalberger** [comment, marked as answer]
`https://www.nuget.org` is the URL of the website, the feed URL for nuget.org is `https://api.nuget.org/v3/index.json`.

Replace

```
.InstallTool(new Uri("nuget:https://www.nuget.org?package=7-Zip.CommandLine"))
```

With

```
.InstallTool(new Uri("nuget:https://api.nuget.org/v3/index.json?package=7-Zip.CommandLine"))
```

**Reply by mandalorianbob** [reply]
Thanks!

**Reply by nick5454** [reply]
I add a nuget.config to my build folders to provide both private and public feeds. Here's my file which works well to preload my private feeds for the build.

`
configuration>
	packageRestore>
		add key="enabled" value="True" />

		
		<add key="automatic" value="True" />
	</packageRestore>

	
	<packageSources>
		<add key="NuGet official package source" value="https://api.nuget.org/v3/index.json" />
		<add key="smarts" value="https://pkgs.dev.azure.com/orgname/_packaging/feedname/nuget/v3/index.json" />
	</packageSources>

	
	<packageSourceCredentials>
		<smarts>
			<add key="Username" value="VSTS" />
			<add key="ClearTextPassword"  value="%NUGET_AUTH_TOKEN%" />
		</smarts>
	</packageSourceCredentials>

	
	<disabledPackageSources />

	
	

	

</configuration>`

The nuget_auth_token is just $(System.AccessToken) so the private organization feeds in azure always have rights. This preloads the feed without needing to specify them
